### ISSUE TITLE
React Native build breaking - 64 bit support

### ISSUE THREAD

**Issue by dbarner1** [realm/realm-js#2684]
Hi folks,

Our react-native xcode build is failing to archive due to realm.   I've searched exhaustively, but haven't found a lead on what could fix this.  It looks like it could be due to a 64 bit support requirement?  Here's the error:

```
Undefined symbols for architecture arm64:
  "realm::util::terminate_internal(std::__1::basic_stringstream<char, std::__1::char_traits<char>, std::__1::allocator<char> >&)", referenced from:
      void realm::util::terminate<unsigned long, unsigned long>(char const*, char const*, long, unsigned long, unsigned long) in libRealm.a(RLMArrayLinkView.o)
      void realm::util::terminate<unsigned long, int>(char const*, char const*, long, unsigned long, int) in libRealm.a(RLMArrayLinkView.o)
      void realm::util::terminate_with_info<int&&gt…
```

## Goals

The build to run successfully

## Expected Results

The build to run successfully.

## Actual Results

The build broke at this error.

## Code Sample

```
Undefined symbols for architecture arm64:
  "realm::util::terminate_internal(std::__1::basic_stringstream<char, std::__1::char_traits<char>, std::__1::allocator<char> >&)", referenced from:
      void realm::util::terminate<unsigned long, unsigned long>(char const*, char const*, long, unsigned long, unsigned long) in libRealm.a(RLMArrayLinkView.o)
      void realm::util::terminate<unsigned long, int>(char const*, char const*, long, unsigned long, int) in libRealm.a(RLMArrayLinkView.o)
      void realm::util::terminate_with_info<int&&gt…
```

## Version of Realm and Tooling

- Realm JS SDK Version: 3.6.0
- Node or React Native: 61.5
- Client OS & Version: Mac 10.14.5
- Which debugger for React Native: None

**Comment by kneth** [comment]
It looks like that linking to done against Realm for 64 bit ARM. Since Macs are Intel machines, something is set up incorrectly. 

How do you install Realm?

**Comment by dbarner1** [comment]
This was run from bitrise.  Here's the full environment:
https://github.com/bitrise-io/bitrise.io/blob/master/system_reports/osx-vs4mac-stable.log

On local install, we have two do 2 pod installs to install realm and run:

cd root-directory
npm install
cd ios
pod install
cd ..
react-native run-ios (it fails)
cd ios
pod install
react-native run-ios (it succeeds)

**Comment by sullyinc** [comment]
@dbarner1 After adding realm to our React Native project we encountered this error too.  Bitrise archives failed with missing symbols for arm64/arm7, archives also failed on a local machine with a clean checkout of the repo. We have the Pods directory checked into source control, with realm 6.0.1 and react-native 0.62.1.

Tried many things but nothing worked. Finally fixed it by deleting  the Pods and node_modules directories and reinstalling. Archive was successful after that. In hindsight, not sure if deleting node_modules was necessary.

```
$ rm -rf node_modules
$ rm -rf ios/Pods
$ yarn install
$ cd ios && pod install && cd ..
$ open ios/[MyApp].xcworkspace/
-> Xcode Archive
```

Hope that helps.

**Comment by sullyinc** [comment]
@dbarner1 We also upgraded our cocoapods, that may have also contributed. 

```
$ pod --version
1.9.2
```

**Comment by dbarner1** [comment]
Thanks @sullyinc

**Comment by sullyinc** [comment]
@dbarner1 I think I spoke too soon, this is still an issue for us. I think it only worked after repeated attempts, like you mentioned earlier.

**Comment by dbarner1** [comment]
@sullyinc bummer, ya.  We ended up sadly moving away from realm for now.  

Still got love for realm though and hope this can get resolved.

**Comment by bkbooth** [comment]
Everytime I add/remove/update dependencies I have to

```
(cd ios && pod install)
yarn run ios
# ☝️ this one fails
(cd ios && pod install)
yarn run ios
```

It's slightly annoying, but it seems to work consistently at least.

**Comment by sullyinc** [comment]
Digging through some builds here, looks like the failing builds/archives are missing parameters for the Xcode link step. Failed builds are missing these params:

```
-L "/Users/user/my-react-app/ios/Pods/../../node_modules/realm/vendor/realm-ios\"
...
-lrealm-ios -lrealm-parser-ios
```

Which would explain the "missing symbols" error. There's some kind of syncing issue between downloading the libs (`node_modules/realm/scripts/download-realm.js`) and setting up the linker params. Realm team, let me know if you have trouble reproducing this.

In the meantime, one workaround is to manually link the vendor libs:

Download and copy to project:
<img width="621" alt="Screen Shot 2020-05-29 at 8 41 08 PM" src="https://user-images.githubusercontent.com/10899740/83316973-0fa5c800-a1ef-11ea-8e3c-64147ce5b68d.png">

Drag and drop into Xcode:
<img width="279" alt="Screen Shot 2020-05-29 at 8 41 41 PM" src="https://user-images.githubusercontent.com/10899740/83316982-1b918a00-a1ef-11ea-9d00-8a0dab0a1294.png">

Verify libs are recognized:
<img width="588" alt="Screen Shot 2020-05-29 at 8 41 56 PM" src="https://user-images.githubusercontent.com/10899740/83316990-28ae7900-a1ef-11ea-90d1-2608aa19d0d4.png">

For me, this fixed our Bitrise/CI builds.

Looking forward to a fix from Realm. Thanks for all the work, looking to use Realm as an upgrade from AsyncStorage.

**Comment by joshua-davis-rose** [comment]
Unfortunately this solution didn't work for me. I'm getting both `Undefined symbols for architecture armv7` and `Undefined symbols for architecture arm64` errors when linking RealmJS on Travis/CI. Frustratingly I can't reproduce the issue on my development machine which is making debugging harder.

Did you need to add anything to your search paths or Other Linker Flags to get this to work?

**Comment by joshua-davis-rose** [comment]
I found the issue, but I don't yet know the solution. On Travis/CI it's not pulling in the realm-ios sources so I end up with a vendor directory listing like this:
```
Traviss-Mac:vendor travis$ ls -al
total 40
drwxr-xr-x   7 travis  staff   224 Jun  5 13:39 .
drwxr-xr-x  20 travis  staff   640 Jun  5 13:40 ..
-rw-r--r--   1 travis  staff     8 Jun  5 13:36 .gitignore
-rw-r--r--   1 travis  staff   248 Jun  5 13:36 .npmignore
drwxr-xr-x   7 travis  staff   224 Jun  5 13:39 GCDWebServer
-rw-r--r--   1 travis  staff  6391 Jun  5 13:36 base64.cpp
-rw-r--r--   1 travis  staff  1788 Jun  5 13:36 base64.hpp
```
Whereas locally, it looks like this:
```
macbook:vendor jrose$ ls -al
total 40
drwxr-xr-x   8 jrose  staff   256 Jun  4 23:44 .
drwxr-xr-x  20 jrose  staff   640 Jun  4 23:43 ..
-rw-r--r--   1 jrose  staff     8 Jun  4 19:23 .gitignore
-rw-r--r--   1 jrose  staff   248 Jun  4 19:23 .npmignore
drwxr-xr-x   7 jrose  staff   224 Jun  4 23:43 GCDWebServer
-rw-r--r--   1 jrose  staff  6391 Jun  4 19:23 base64.cpp
-rw-r--r--   1 jrose  staff  1788 Jun  4 19:23 base64.hpp
drwxr-xr-x  26 jrose  staff   832 Jun  4 23:44 realm-ios
```

**Comment by sullyinc** [comment]
@joshua-davis-rose Very true, it's especially frustrating to debug linker problems when builds take 25 minutes, like for React Native apps.

Ah you're correct, I missed the final step. Looks like I had to add the `realm-lib-patch` directory to Library Search Paths:

<img width="753" alt="Screen Shot 2020-06-05 at 9 25 49 AM" src="https://user-images.githubusercontent.com/10899740/83887895-b23dd980-a70e-11ea-8238-062d3628101a.png">

Hope that fixes it.
 
I don't know the real solution either, but I also found it's only pulling in the `realm-ios` folder and libraries only after building the app and running `pod install` a few times.

**Comment by joshua-davis-rose** [comment]
Unfortunately that still didn't work for me, but I think I sorted out what's going on. For some reason the `script_phases` of the podspec isn't getting executed during `pod install` and only during pre-compile (which I see in the logs), but I think it's too late at that point. I found a workaround by adding this to my package.json scripts:
```
"scripts": {
        "postinstall": "node node_modules/realm/scripts/download-realm.js ios --sync"
}
```
I was finally able to build on Travis/CI

**Comment by kneth** [comment]
The original post hints that the app is using both Realm Cocoa and Realm JavaScript. In the past, we haven't had good support for that but adding better Cocoapod support should have improved it. So the question to you all is: do you still experience the issue with a recent version of Realm JavaScript and Realm Cocoa?

**Comment by kneth** [comment]
I am closing the issue due to lack of activity. Please either reopen or create new issue if you still observe it.
